lib/ble.toit

```diff
@@ -171,339 +116,389 @@ class RemoteDevice:
-A client connected to a remote device.
+A remote connected device.
 */
-class Client:
-  device/Device
+class RemoteConnectedDevice extends Resource_:
+  manager/CentralManager
   /**
   The address of the remote device the client is connected to.
   */
-  address/Address
+  address/any
 
-  gatt_ := ?
-  resource_state_/monitor.ResourceState_
+  discovered_services/List := []
 
-  constructor .device .address:
-    gatt_ = ble_get_gatt_ device.resource_group_
-    resource_state_ = monitor.ResourceState_ device.resource_group_ gatt_
-    ble_connect_ device.resource_group_ address.raw_ gatt_
+  constructor .manager .address:
+    device := ble_connect_ manager.resource_ address
+    super manager.adapter.resource_group_ device --auto_release
     state := resource_state_.wait_for_state CONNECTED_EVENT_ | CONNECT_FAILED_EVENT_
     if state & CONNECT_FAILED_EVENT_ != 0:
+      close_
       throw "BLE connection failed"
       // TODO Possible leak of the gatt_ resource on connection failures
 
   /**
-  Reads a remote service by looking up the given $service_uuid on the remote device.
-  */
-  read_service service_uuid/uuid_pkg.Uuid -> RemoteService:
-    ble_request_service_ gatt_ service_uuid.to_byte_array
-    wait_for_done_
-    result := ble_request_result_ gatt_
-    return RemoteService this service_uuid result
+  Discovers remote services by looking up the given $service_uuids on the remote device
 
-  wait_for_done_:
-    state := resource_state_.wait_for_state STARTED_EVENT_
-    resource_state_.clear_state STARTED_EVENT_
+  If $service_uuids is empty all services for the device is discovered.
 
-/**
-Contains definitions of the services exposed by the BLE server.
-*/
-class ServerConfiguration:
-  resource_group_ := null
-  services/List := []
+  Note: Some platforms only support an empty list or a list of size 1. If the platform is limited, this method
+  will throw INVALID_ARGUMENT.
+  */
+  discover_services service_uuids/List=[] -> List:
+    resource_state_.clear_state SERVICES_DISCOVERED_EVENT_
+    raw_service_uuids := service_uuids.map: | uuid/BleUUID | uuid.encode_for_platform_
+    ble_discover_services_ resource_ (Array_.ensure raw_service_uuids)
+    state := wait_for_state_with_gc_ SERVICES_DISCOVERED_EVENT_
+                                   | DISCONNECTED_EVENT_
+                                   | DISCOVERY_OPERATION_FAILED_
+    if state & DISCONNECTED_EVENT_ != 0:
+      throw "BLE disconnected"
+    else if state & DISCOVERY_OPERATION_FAILED_ != 0:
+      throw_error_
+
+    discovered_services.add_all
+        List.from
+            (ble_discover_services_result_ resource_).map:
+              RemoteService this (BleUUID it[0]) it[1]
+
+    return order_attributes_ service_uuids discovered_services
 
-  constructor:
-    resource_group_ = ble_server_configuration_init_
-    add_finalizer this:: ble_server_configuration_dispose_ resource_group_
+  /**
+  Disconnects from the remote device.
+  */
+  disconnect:
```

Comment:
Since we don't have a `connect` function, I would go with `close` here.

If we ever get some "auto-close" functionality, it would work here too. (Not that anything is planned).

Reply:
CentralManager.connect?

Reply:
True. But it's not a "connect", "disconnect", "connect", dance, where you can do this multiple times.

Also, for sockets we do a connect, and then call `close` on the object.